File: leetcode/python/P1457.py
"""
Problem: https://leetcode.com/problems/pseudo-palindromic-paths-in-a-binary-tree/
Solution: https://leetcode.com/submissions/detail/799450744/
"""
from typing import Optional
from models import TreeNode

class Solution:
  def pseudoPalindromicPaths(self, root: Optional[TreeNode]) -> int:
    arr = dict()
    return self.traverse(root, arr)

  @staticmethod
  def traverse(node: Optional[TreeNode], arr: dict[int, int]) -> int:
    if node is None:
      return 0
    
    if (arr.get(node.val) is None):
      arr[node.val] = 0
    
    arr[node.val] += 1

    val = 0
    if node.left is None and node.right is None:
      val = 1 if Solution.is_palindrome(arr) else 0
    else:
      val = Solution.traverse(node.left, arr) + Solution.traverse(node.right, arr) 

    arr[node.val] -= 1
    return val

  @staticmethod
  def is_palindrome(arr) -> bool:
    odds = 0
    for k, v in arr.items():
      if (v % 2 == 1):
        odds += 1
    return odds <= 1
        

File: leetcode/c#/Problems/1400/P1457.cs
namespace LeetCode.Naive.Problems;

/// <summary>
///    Problem: https://leetcode.com/problems/pseudo-palindromic-paths-in-a-binary-tree/
///    Submission: https://leetcode.com/submissions/detail/344558223/
/// </summary>
internal class P1457
{
  public class Solution
  {
    public int PseudoPalindromicPaths(TreeNode root)
    {
      var list = new List<int>();
      return Traverse(root, list);
    }

    private int Traverse(TreeNode node, List<int> list)
    {
      if (node == null)
        return 0;

      list.Add(node.val);

      var val = 0;

      if (node.left == null && node.right == null)
        val = IsPalindrome(list) ? 1 : 0;
      else
        val = Traverse(node.left, list) + Traverse(node.right, list);

      list.RemoveAt(list.Count - 1);

      return val;
    }

    private bool IsPalindrome(List<int> list)
    {
      var odds = list.GroupBy(d => d).ToDictionary(x => x.Key, x => x.Count())
        .Where(x => x.Value % 2 == 1)
        .Count();

      return odds <= 1;
    }
  }
}

/// <summary>
///    Problem: https://leetcode.com/problems/pseudo-palindromic-paths-in-a-binary-tree/
///    Submission: https://leetcode.com/submissions/detail/799425473/
///    Optimized
/// </summary>
internal class P1457_2
{
  public class Solution
  {
    public int PseudoPalindromicPaths(TreeNode root)
    {
      var list = new int[10];
      return Traverse(root, list);
    }

    private int Traverse(TreeNode node, int[] list)
    {
      if (node == null)
        return 0;

      list[node.val]++;

      int val;
      if (node.left == null && node.right == null)
      {
        val = IsPalindrome(list) ? 1 : 0;
      }
      else
      {
        val = Traverse(node.left, list) + Traverse(node.right, list);
      }

      list[node.val]--;
      return val;
    }

    private bool IsPalindrome(int[] list)
    {
      var odds = 0;
      for (int i = 0; i < list.Length; i++)
      {
        if (list[i] % 2 == 1)
          odds++;
      }

      return odds <= 1;
    }
  }

}
